### PR TITLE
feat: support scrape_protocols for GlobalConfig and ScrapeConfig

### PR DESCRIPTION
**Pull request by dongjiang1989**
## Description

_Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request.
If it fixes a bug or resolves a feature request, be sure to link to that issue._

ref: https://github.com/prometheus/prometheus/pull/12738

support scrape_protocols for GlobalConfig and ScrapeConfig

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
new option in GlobalConfig and ScrapeConfig: "scrape_protocols"
```

### CODE REVIEW

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 1]
```diff
@@ -32,6 +32,15 @@ const (
 	PrometheusKindKey = "prometheus"
 )
 
+// ScrapeProtocol represents a protocol used by Prometheus for scraping metrics.
+// Supported values are:
```
Comment (slashpai) [inline pkg/apis/monitoring/v1/prometheus_types.go:36]: Do we want to specify this if validation is enabled?

**Review by slashpai — COMMENTED** [review]
just a nit, lgtm otherwise

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 1 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:36]: By adding the kubebuilder enumeration validation, and the value can be null using the prometheus default.
```
// +kubebuilder:validation:Enum=PrometheusProto;OpenMetricsText0.0.1;OpenMetricsText1.0.0;PrometheusText0.0.4
```

File: pkg/apis/monitoring/v1/prometheus_types.go [hunk 2]
```diff
@@ -252,6 +261,17 @@ type CommonPrometheusFields struct {
 	// Number of seconds to wait until a scrape request times out.
 	ScrapeTimeout Duration `json:"scrapeTimeout,omitempty"`
 
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1/prometheus_types.go:265]: (nit)
```suggestion
	// The protocols to negotiate during a scrape. It tells clients the
	// protocols supported by Prometheus in order of preference (from most to least preferred).
```

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go [hunk 3]
```diff
@@ -157,6 +157,16 @@ type ScrapeConfigSpec struct {
 	// ScrapeTimeout is the number of seconds to wait until a scrape request times out.
 	// +optional
 	ScrapeTimeout *v1.Duration `json:"scrapeTimeout,omitempty"`
+	// The protocols to negotiate during a scrape. It tells clients what
+	// protocol are accepted by Prometheus and with what weight (most wanted is first).
+	//
+	// If unset, Prometheus uses its default value: ["OpenMetricsText1.0.0","OpenMetricsText0.0.1","PrometheusText0.0.4"]
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:163]: I'd remove the Prometheus default + same remark than above.

**Review by ArthurSens — APPROVED** [review]
Besides simon's comments, LGTM

This could also be added to Pod/Service monitors, but let's do it in a separate PR to keep the small scope here

**Comment by dongjiang1989** [comment]
> Besides simon's comments, LGTM
> 
> This could also be added to Pod/Service monitors, but let's do it in a separate PR to keep the small scope here

Thanks. Got it.

File: pkg/apis/monitoring/v1/prometheus_types.go (hunk 2 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1/prometheus_types.go:265]: Thanks for your review.  
Update Docs done.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go (hunk 3 above)
Comment (dongjiang1989) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:163]: Thanks. remove remark done.

File: pkg/prometheus/promcfg.go [hunk 4]
```diff
@@ -2064,6 +2065,16 @@ func (cg *ConfigGenerator) appendScrapeIntervals(slice yaml.MapSlice) yaml.MapSl
...
 }
 
+func (cg *ConfigGenerator) appendScrapeProtocols(slice yaml.MapSlice) yaml.MapSlice {
+	cpf := cg.prom.GetCommonPrometheusFields()
+
+	if len(cpf.ScrapeProtocols) > 0 {
+		slice = cg.WithMinimumVersion("2.49.0").AppendMapItem(slice, "scrape_protocols", cpf.ScrapeProtocols)
+	}
+
+	return slice
```
Comment (simonpasquier) [inline pkg/prometheus/promcfg.go:2075]: (nit) a bit more compact
```suggestion
	cpf := cg.prom.GetCommonPrometheusFields()

	if len(cpf.ScrapeProtocols) == 0 {
		return slice
	}

	return cg.WithMinimumVersion("2.49.0").AppendMapItem(slice, "scrape_protocols", cpf.ScrapeProtocols)
```

**Review by simonpasquier — COMMENTED** [review]
one small nit but it's only stylistic.

**Comment by dongjiang1989** [comment]
> one small nit but it's only stylistic.

Thanks. Done

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**